hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/Contexts.java

```diff
@@ -108,26 +131,76 @@ public ILogger logger() {
         public ProcessingGuarantee processingGuarantee() {
             return processingGuarantee;
         }
+
     }
 
     static class ProcSupplierCtx extends MetaSupplierCtx implements ProcessorSupplier.Context {
 
         private final int memberIndex;
+        private final ConcurrentMap<String, File> tempDirectories;
 
         @SuppressWarnings("checkstyle:ParameterNumber")
         ProcSupplierCtx(
-                JetInstance jetInstance, long jobId, long executionId, JobConfig jobConfig, ILogger logger,
-                String vertexName, int localParallelism, int totalParallelism, int memberIndex, int memberCount,
-                ProcessingGuarantee processingGuarantee) {
+                JetInstance jetInstance,
+                long jobId,
+                long executionId,
+                JobConfig jobConfig,
+                ILogger logger,
+                String vertexName,
+                int localParallelism,
+                int totalParallelism,
+                int memberIndex,
+                int memberCount,
+                ProcessingGuarantee processingGuarantee,
+                ConcurrentMap<String, File> tempDirectories
+        ) {
             super(jetInstance, jobId, executionId, jobConfig, logger, vertexName, localParallelism, totalParallelism,
                     memberCount, processingGuarantee);
             this.memberIndex = memberIndex;
+            this.tempDirectories = tempDirectories;
         }
 
         @Override
         public int memberIndex() {
             return memberIndex;
         }
+
+        @Nonnull @Override
+        public File attachedDirectory(@Nonnull String id) {
+            Preconditions.checkHasText(id, "id cannot be null or empty");
+            return tempDirectories.computeIfAbsent(id, x -> extractFileToDisk(id));
```

Comment:
`computIfAbsent` method doesn't guarantee that the value object generated by the mapping function will end up as a value in the map. It uses `putIfAbsent` internally and if that fails it fetches the actual value from the map. The value object created with the mapping function will be left on the disk untracked.

Reply:
That's not true, CHM has a special implementation of `computeIfAbsent` which is atomic.

Reply:
I've missed that, I was looking at the default implementation on the ConcurrentMap. Good to know that.

Reply:
We should then use CHM for the field, not just CM.